### PR TITLE
Include dataset and class descriptions in CSV for import/export

### PR DESCRIPTION
**Pull request by aidanlw17**
### Import/Export csv w/ description support

This pull request addresses the issues noted in ticket AB-387
https://tickets.metabrainz.org/projects/AB/issues/AB-387?filter=allopenissues

##### Import:
- alter `_parse_dataset_csv(file)` in `webserver/views/datasets.py` to
  handle dataset and class descriptions on any line of csv file
- csv line format for dataset description: 
    `description,description of dataset dataset`
- csv line format for class description: 
    `description:class1,description of class 1`
- `_parse_dataset_csv(file)` returns `[dataset_description, classes]`
  where dataset_description is description of entire dataset and classes
  is a list of all class dictionaries with structure:
```    
    {'recordings': [mbid1,...,mbidn], 
    'description': 'class description', 
    'name': 'class_name'}
```
- small change to `db/dataset.py` function
  `create_from_dict(dictionary,author_id)` to support class and dataset
  dictionaries containing a description

##### Export:
- in `webserver/views/datasets.py` function
  `_convert_dataset_to_csv_stringio(dataset)` adds dataset description and 
  class descriptions to top of csv if they exist
- format:
```
    description,dataset description
    description:class1,description of class one 
    description:class2,description of class two 
    ... 
    mbid,classname
    ... 
```

Create unit test `test_parse_dataset_csv` in `webserver/views/test:
- add directory `webserver/views/test_data` with file `test_db.csv`
- add `test_data` path to `webserver/testing.py`

Altered unit test for `test_dataset_to_csv`:
- add support for class and dataset descriptions

### CODE REVIEW

**Comment by alastair** [comment]
@brainzbot add to whitelist

File: webserver/views/datasets.py [hunk 1]
```diff
@@ -122,6 +122,19 @@ def _convert_dataset_to_csv_stringio(dataset):
     fp = StringIO.StringIO()
     writer = csv.writer(fp)
 
+    if dataset["description"] is not None:
+        #check if description is empty string
+        if dataset["description"] != "":
```
Comment (alastair) [inline webserver/views/datasets.py]: You don't need to do this check twice here, you can do `if not dataset["descripton"]`, which will check against it being None and empty.

File: webserver/views/datasets.py [hunk 2]
```diff
@@ -286,10 +299,11 @@ def create_service():
 def import_csv():
     form = forms.DatasetCSVImportForm()
     if form.validate_on_submit():
+        [description, classes] = _parse_dataset_csv(request.files[form.file.name])
         dataset_dict = {
             "name": form.name.data,
-            "description": form.description.data,
-            "classes": _parse_dataset_csv(request.files[form.file.name]),
+            "description": description if not None else form.description.data,
```
Comment (alastair) [inline webserver/views/datasets.py]: did you mean to say `description if description not None`? You can also say `description if description else form...`

File: webserver/views/datasets.py (hunk 1 above)
Comment (aidanlw17) [reply webserver/views/datasets.py]: Noted. I'll make this change and update this.

**Comment by aidanlw17** [comment]
Thanks @alastair. I will look into using git rebase and can separate these changes. 

I was meaning to inquire about this when I realized the issue with committing to CONTRIBUTIONS.md. I've forked the acoustic-brainz server and was making all of my changes on the master branch so I made this pull request with that branch as well, which seemed to mean that any changes I committed to my fork's master branch were automatically added to this pull request. 

In the future, does it work better to create a separate branch that is up to date with metabrainz/acoustic-server (rather than ahead, like the master branch for my fork) for each feature/issue I'm working on? Then create individual pull requests on each of these branches and merge them to the master branch of my fork only after their changes have been accepted to metabrainz/acoustic-server?

Thanks again, the feedback is greatly appreciated and I apologize for the confusing commits in this PR!

**Comment by alastair** [comment]
Absolutely, it's always best to create a new branch from the most recent version of master in order to make your changes. This means that if you want to you're able to make multiple changes at once (e.g. this one, one for CONTRIBUTING, and one for fixing .gitignore).
Take a look at github's documentation about "github flow": https://guides.github.com/introduction/flow/ (You could also add this to the contributing doc - I thought it was adequately described in the github forking guide under the "Git workflow" section, but this doesn't seem to be the case).
Take a look at the contributing doc for Listenbrainz to see some suggested changes that we could add: https://github.com/metabrainz/listenbrainz-server/blob/master/.github/CONTRIBUTING.md#git-workflow

**Comment by aidanlw17** [comment]
@alastair thank you for clearing that up! I think that I better understand the workflow with branching and I can make these changes in a clearer way by doing so. I'll certainly see what I can add from Listenbrainz to CONTRIBUTING about this workflow alongside the other things that I was planning to add. 

For the .gitignore fix, I added that because the package-lock.json and the config.py that I created for my local dev environment were committed to my fork but didn't exist in the acoustic-brainz sever repo. Should those two things actually be added to the .gitignore on the master acoustic-brainz as well or is this something that contributors to acousticbrainz generally manage on their own?

**Pull request closed without merging**